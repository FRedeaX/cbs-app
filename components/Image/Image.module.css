.placeholder {
  position: relative;
  overflow: hidden;
}

.blur {
  opacity: 0.95;
  filter: blur(var(--image-filter-blur, 10px));
}

.blur,
.skeleton {
  position: absolute;
  left: 0;
}

.blur,
.image {
  max-width: 100%;
  height: auto;
}

.placeholder .blur {
  object-fit: cover;
}

.placeholder .skeleton {
  position: absolute;
  transform: none;
  background-color: transparent;
}

.placeholder .skeleton::after {
  background: linear-gradient(
    90deg,
    transparent,
    rgba(0, 0, 0, 8%),
    transparent
  );
}

.placeholder,
.image {
  border-radius: var(--image-border-radius, var(--g-image-border-radius));
}

.image {
  display: flex;

  /* stylelint-disable-next-line font-family-no-missing-generic-family-keyword */
  font: 0 / 0 a;
  visibility: hidden; /* Позволяет не показывать битые картинки при ошибках загрузки */
  border: 0;
}

.image_loaded {
  visibility: visible;
}

.image_loading {
  opacity: 0;
}

.image_animated {
  animation: image-fade-in 0.4s ease forwards;
}

@keyframes image-fade-in {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}
