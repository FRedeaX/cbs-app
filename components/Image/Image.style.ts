import { BoxProps } from "@mui/material";

export const sxRoot: BoxProps["sx"] = {
  position: "relative",
  display: "flex",
  width: "100%",
  height: "100%",
};
