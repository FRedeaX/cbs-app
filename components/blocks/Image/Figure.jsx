import { gql, useQuery } from "@apollo/client";
import classNames from "classnames";
import { useEffect, useRef, useState } from "react";

import { createMarkup, delay } from "../../../helpers";
import {
  GET_OVERLAY_FRAGMENT,
  overlayVar,
} from "../../../store/variables/overlay";
import { SCROLLY_FRAGMENT } from "../../../store/variables/scrollY";
import Button from "../../UI/Button/Button";
import Icon from "../../UI/Icon/Icon";
import classes from "./Image.module.css";

export const imageBlockGQL = {
  fragments: gql`
    fragment imageBlockGQL on CoreImageBlock {
      ... on CoreImageBlock {
        name
        attributes {
          ... on CoreImageBlockAttributes {
            id
            align
            caption
            alt
            className
            url
          }
        }
      }
    }
  `,
};

export const Figure = ({
  caption,
  align,
  className,
  url,
  isImageZoom = true,
  children,
}) => {
  const { data: state } = useQuery(gql`
    query {
      ${GET_OVERLAY_FRAGMENT}
      ${SCROLLY_FRAGMENT}
    }
  `);
  const figureRef = useRef();
  const [isZoom, setZoom] = useState();

  const positionScrollYRefVar = useRef(0);
  useEffect(() => {
    if (
      (!state.overlay.isOpen ||
        state.scrollY !== positionScrollYRefVar.current) &&
      isZoom
    ) {
      if (state.scrollY !== positionScrollYRefVar.current)
        overlayVar({ isOpen: false, color: "var(--bg-white-95)" });

      setZoom(false);
      delay(250).then(() => {
        figureRef.current.style.zIndex = "";
      });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [state?.overlay.isOpen, state?.scrollY]);

  const getTransform = () => {
    const figure = figureRef.current;
    const image = figure.children[0].children[1];

    const { clientWidth } = document.documentElement;
    const { clientHeight } = document.documentElement;
    // const pxWidth = image.x + image.clientWidth;
    // const pxHeight = image.y + image.clientHeight;
    const { devicePixelRatio } = window;

    const imageX = image.x / devicePixelRatio + image.width / 2;
    const imageY = image.y / devicePixelRatio + image.height / 2;
    const centerX = clientWidth / 2 - imageX;
    const centerY = clientHeight / 2 - imageY;

    const scaleX = clientWidth / image.width;
    const scaleY = clientHeight / image.height;

    return `translate(${centerX}px, ${centerY}px) scale(${Math.min(
      scaleX,
      scaleY,
      1.6,
    )
      .toString()
      .substr(0, 3)})`;
  };
  const hendleZoom = (event) => {
    event.stopPropagation();
    if (isZoom) {
      setZoom(false);
      overlayVar({ isOpen: false, color: "var(--bg-white-95)" });
      delay(250).then(() => {
        figureRef.current.style.zIndex = "";
      });
    } else {
      setZoom(true);
      overlayVar({
        isOpen: true,
        opacity: 100,
        color: "white",
        isOverflow: false,
      });
      positionScrollYRefVar.current = state?.scrollY;
    }
  };

  // if (align === "left") align = "flex-start";
  // else if (align === "right") align = "flex-end";
  return (
    <div
      className={classNames(
        classes.block,
        {
          [classes[`block_align_${align}`]]: align !== "",
        },
        className,
      )}>
      <figure
        ref={figureRef}
        onClick={isImageZoom ? hendleZoom : null}
        onKeyPress={isImageZoom ? hendleZoom : null}
        // eslint-disable-next-line jsx-a11y/no-noninteractive-element-to-interactive-role
        role="button"
        tabIndex="0"
        style={{
          transform: isZoom ? getTransform() : "",
          zIndex: isZoom ? "4" : delay(250).then(""),
          cursor: isZoom && "default",
        }}
        className={classNames(classes.figure, {
          [classes.figure_zoom_true]: isZoom === true,
        })}>
        {children}
        {caption && (
          <figcaption
            className={classes.caption}
            dangerouslySetInnerHTML={createMarkup(caption)}
          />
        )}
        <div className={classes.controls}>
          <Button
            className={classes.button_download}
            href={url}
            view="download"
            icon={<Icon type="download" isGlyph side={false} />}
          />
        </div>
      </figure>
    </div>
  );
};
