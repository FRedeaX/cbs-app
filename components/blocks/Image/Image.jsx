import { gql, useQuery } from "@apollo/client";
import classNames from "classnames";
// import Image from "next/image";
import { default as ImageNext } from "next/image";
import { useRef, useState } from "react";
import { useEffect } from "react/cjs/react.development";
import { delay } from "~/helpers/delay";
import { GET_OVERLAY_FRAGMENT, overlayVar } from "~/store/variables/overlay";
import { SCROLLY_FRAGMENT } from "~/store/variables/scrollY";
import Button from "../../UI/Button/Button";
import { Icon } from "../../UI/Icon/Icon";
import classes from "./Image.module.css";

export const imageBlockGQL = {
  fragments: gql`
    fragment imageBlockGQL on CoreImageBlock {
      ... on CoreImageBlock {
        name
        attributes {
          ... on CoreImageBlockAttributes {
            id
            align
            caption
            alt
            className
            url
          }
        }
      }
    }
  `,
};

export const Image = ({
  id,
  caption,
  alt,
  align,
  className,
  url,
  width,
  height,
  isImageZoom = true,
}) => {
  const { data: state } = useQuery(gql`
    query {
      ${GET_OVERLAY_FRAGMENT}
      ${SCROLLY_FRAGMENT}
    }
  `);
  const figureRef = useRef();
  const [isZoom, setZoom] = useState();

  const positionScrollYRefVar = useRef(0);
  useEffect(() => {
    if (
      (!state.overlay.isOpen ||
        state.scrollY !== positionScrollYRefVar.current) &&
      isZoom
    ) {
      if (state.scrollY !== positionScrollYRefVar.current)
        overlayVar({ isOpen: false, color: "white" });

      setZoom(false);
      delay(250).then(() => (figureRef.current.style.zIndex = ""));
    }
  }, [state?.overlay.isOpen, state?.scrollY]);

  const getTransform = () => {
    const figure = figureRef.current;
    const image = figure.children[0].children[1];

    const clientWidth = document.documentElement.clientWidth;
    const clientHeight = document.documentElement.clientHeight;
    const imageX = image.x + image.clientWidth / 2;
    const imageY = image.y + image.clientHeight / 2;
    const centerX = clientWidth / 2 - imageX;
    const centerY = clientHeight / 2 - imageY;
    let scale = (
      height < width || clientHeight > clientWidth
        ? clientWidth / image.clientWidth
        : clientHeight / image.clientHeight
    )
      .toString()
      .substr(0, 3);

    if (scale > 1.6) scale = 1.6;
    return `translate(${centerX}px, ${centerY}px) scale(${scale})`;
  };
  const hendleZoom = (event) => {
    event.stopPropagation();
    if (isZoom) {
      setZoom(false);
      overlayVar({ isOpen: false, color: "white" });
      delay(250).then(() => (figureRef.current.style.zIndex = ""));
    } else {
      setZoom(true);
      overlayVar({
        isOpen: true,
        opacity: 100,
        color: "white",
        isOverflow: false,
      });
      positionScrollYRefVar.current = state?.scrollY;
    }
  };

  // if (align === "left") align = "flex-start";
  // else if (align === "right") align = "flex-end";
  return (
    <div
      className={classNames(
        classes.block,
        {
          [classes[`block_align_${align}`]]: align !== "",
        },
        className
      )}
    >
      <figure
        key={id}
        ref={figureRef}
        onClick={isImageZoom ? hendleZoom : null}
        style={{
          transform: isZoom ? getTransform() : "",
          zIndex: isZoom ? "4" : delay(0).then(""),
          cursor: isZoom && "default",
        }}
        className={classes.figure}
      >
        <ImageNext
          src={url}
          alt={alt}
          // layout="fill"
          width={width}
          height={height}
          // blurDataURL={base64}
          // placeholder="blur"
          // loading={"lazy"}
          // layout="responsive"
          // objectFit="cover"
          // className={className}
        />
        {caption && (
          <figcaption className={classes.caption}>{caption}</figcaption>
        )}
        <div className={classes.controls}>
          <Button
            className={classes["button_download"]}
            href={url}
            view="download"
            icon={<Icon type={"download"} isGlyph={true} side={false} />}
          />
        </div>
      </figure>
    </div>
  );
};
