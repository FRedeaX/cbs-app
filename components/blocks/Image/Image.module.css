.block {
  --button-border-width: 0;
  --carousel-button-visibility: 0.3;
  --icon-fill: #fff;

  display: flex;
  height: 100%;
  margin: 6px 0 20px;
}

.block:first-child {
  margin-top: 0;
}

.block:last-child {
  margin-bottom: 0;
}

.block_align_left {
  justify-content: flex-start;
}

.block_align_center {
  justify-content: center;
}

.block_align_right {
  justify-content: flex-end;
}

.figure {
  position: relative;
  width: 100%;
  height: 100%;
  margin: 0;

  /* min-width: 100%;
  padding-top: 56.25%; */

  /* overflow: hidden;
  border-radius: 12px; */
  cursor: pointer;

  /* transition: transform 0.25s linear; */
  transition: transform 300ms cubic-bezier(0.1, 0.82, 0.25, 1) 0s;
}

.figure_height > span {
  height: 100% !important;
}

.figure_zoom_true img {
  border-radius: 12px !important;
}

.image {
  border-radius: 12px;
}

.caption {
  position: absolute;
  bottom: 0;
  width: 100%;

  /* max-height: 100%; */
  padding: 40px 10px 9px;
  color: #fff;
  font-size: 14px;
  line-height: 1.2;
  text-align: center;
  background: linear-gradient(
    0deg,
    rgba(0, 0, 0, 70%),
    rgba(0, 0, 0, 30%) 70%,
    transparent
  );
}

.controls {
  position: absolute;
  right: 0;
  bottom: 0;
  z-index: 1;

  /* background-color: white;
  border-radius: 0 12px 0 12px;
  opacity: 0.75; */
}

.button {
  width: 40px;
  height: 40px;
  background-color: white;
  border-radius: 50%;
  box-shadow: 0 0 20px -15px;
}

.button_download {
  width: 44px;
  height: 44px;
  opacity: 0;
  transition: opacity 0.15s linear;
}

.figure:hover .button_download {
  opacity: 0.5;
}

.controls:hover .button_download {
  opacity: 1;
}

.button_download_fill {
  --icon-fill: #fff;
}
