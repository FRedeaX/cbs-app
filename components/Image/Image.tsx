"use client";

/* eslint-disable @next/next/no-img-element */
/* eslint-disable react/jsx-props-no-spreading */
import { Fade, Skeleton } from "@mui/material";
import classNames from "classnames";
import NextImage, { ImageProps as NextImageProps } from "next/image";
import { FC, useCallback, useState, useRef, useEffect } from "react";

import { Nullable } from "../../helpers/typings/utility-types";

import classes from "./Image.module.css";
import { sxRoot } from "./Image.style";
import { getImageBlurSvg } from "./utils/getImageBlurSvg";
import { isRendered } from "./utils/isRendered";
import { ref } from "yup";

const ANIMATION_DELAY_MS = 50;

export type ImageProps = {
  /**
   * Заполнитель для отображения во время загрузки картинки.
   */
  blurDataURL?: Nullable<string>;
  /**
   * Ширина изображения.
   */
  width: number;
  /**
   * Высота изображения.
   */
  height: number;
} & Omit<NextImageProps, "placeholder" | "blurDataURL" | "width" | "height">;

export const Image: FC<ImageProps> = ({
  blurDataURL,
  alt,
  className,
  width,
  height,
  src,
  style,
  ...props
}) => {
  const [isLoaded, setLoaded] = useState(false);
  const [needAnimation, setNeedAnimation] = useState(false);
  const [canRemovePlaceholder, setCanRemovePlaceholder] = useState(false);

  const imageRef = useRef<Nullable<HTMLImageElement>>(null);

  useEffect(() => {
    const ref = imageRef.current;
    const timer = setTimeout(() => {
      if (!isLoaded && ref && !isRendered(ref)) {
        setNeedAnimation(true);
      }
    }, ANIMATION_DELAY_MS);

    return () => {
      clearTimeout(timer);
    };
  }, [isLoaded, setNeedAnimation]);

  const handleOnAnimationEnd = useCallback(() => {
    setCanRemovePlaceholder(true);
  }, []);

  const handleOnLoad = useCallback(() => {
    setLoaded(true);
  }, []);

  const cnPlaceholder = classNames(classes.placeholder, className, {
    [classes.placeholder_animated]: needAnimation && isLoaded,
  });

  return (
    <div ref={imageRef} sx={sxRoot}>
      <NextImage
        ref={imageRef}
        alt={alt}
        aria-hidden={!alt}
        src={src}
        style={style}
        className={classNames(classes.image, className)}
        width={width}
        height={height}
        onLoad={handleOnLoad}
        placeholder={blurDataURL ? "blur" : "empty"}
        blurDataURL={blurDataURL ?? undefined}
        {...props}
      />
      {blurDataURL && needAnimation && !canRemovePlaceholder && (
        <Fade in={needAnimation && !isLoaded} easing="ease" timeout={600}>
          <div
            data-testid="image-placeholder"
            className={cnPlaceholder}
            onAnimationEnd={handleOnAnimationEnd}>
            <div className={classes.backdrop} />
            <img
              alt=""
              aria-hidden
              src={`data:image/svg+xml;charset=utf-8,${getImageBlurSvg({
                width,
                height,
                blurDataURL,
                objectFit: style?.objectFit,
              })}`}
              className={classes.blur}
            />
            <Skeleton className={classes.skeleton} animation="wave" />
          </div>
        </Fade>
      )}
    </div>
  );
};

Image.displayName = "Image";
