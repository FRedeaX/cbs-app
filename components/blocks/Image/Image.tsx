import classNames from "classnames";
import { FC, HTMLAttributes } from "react";

import { useToggle } from "../../../helpers/frontend/hooks";
import { Figure, FigureFigcaption } from "../../Figure";
import {
  Image as AnimatedImage,
  ImageProps as IAnimatedImageProps,
} from "../../Image/Image";
import {
  ImageViewer,
  ImageViewerBody,
  ImageViewerButton,
  ImageViewerFigure,
  ImageViewerHeader,
  ImageViewerZoom,
} from "../../ImageViewer";
import { AlignImage } from "../utils/types";

import classes from "./Image.module.css";

type ImagePropsExtends = "alt" | "src" | "width" | "height" | "blurDataURL";

export type ImageProps = {
  /**
   * Расположение изображения по оси `x`.
   */
  align?: AlignImage;
  /**
   * HTML-якорь.
   */
  anchor?: string;
  /**
   * Подпись.
   */
  caption?: string;
  /**
   * Дополнительный класс.
   */
  className?: string | classNames.ArgumentArray;
  AnimatedImageProps?: Omit<IAnimatedImageProps, ImagePropsExtends>;
} & Pick<IAnimatedImageProps, ImagePropsExtends> &
  Omit<HTMLAttributes<HTMLDivElement>, "className">;

export const Image: FC<ImageProps> = ({
  align,
  anchor,
  caption,
  alt,
  src,
  width,
  height,
  blurDataURL,
  className,
  AnimatedImageProps,
  ...props
}) => {
  const [isOpen, setToggle] = useToggle(false);
  const captionOrAlt = caption || alt;

  return (
    <div
      id={anchor || undefined}
      className={classNames(classes.root, className)}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}>
      <ImageViewer open={isOpen} onClose={setToggle}>
        <ImageViewerHeader onClose={setToggle} />
        <ImageViewerBody>
          <ImageViewerFigure
            className={classes.imageWrapper}
            figcaptionText={captionOrAlt}>
            <ImageViewerZoom>
              <AnimatedImage
                alt={alt}
                src={src}
                width={width}
                height={height}
                sizes="100vw"
                quality={85}
                loading="lazy"
                blurDataURL={blurDataURL}
              />
            </ImageViewerZoom>
          </ImageViewerFigure>
        </ImageViewerBody>
      </ImageViewer>

      <div className={classes[`align_${align}`]}>
        <Figure>
          <ImageViewerButton onClick={setToggle}>
            <AnimatedImage
              // eslint-disable-next-line react/jsx-props-no-spreading
              {...AnimatedImageProps}
              alt={alt}
              src={src}
              width={width}
              height={height}
              sizes="100vw"
              className={classes.image}
              loading="lazy"
              blurDataURL={blurDataURL}
            />
          </ImageViewerButton>
          {captionOrAlt && (
            <FigureFigcaption text={captionOrAlt} position="outside" />
          )}
        </Figure>
      </div>
    </div>
  );
};
