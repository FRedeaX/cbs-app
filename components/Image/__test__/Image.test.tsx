import { act } from "react";
import { fireEvent, render, screen } from "@testing-library/react";

import { Image } from "../Image";

const EMPTY_IMAGE_SRC =
  "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7";

describe("<Image />", () => {
  test("должен вернуть полный шаблон компонента (snapshot)", () => {
    const { asFragment } = render(
      <Image alt="" src={EMPTY_IMAGE_SRC} width={40} height={40} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("должен установить атрибут alt", () => {
    render(
      <Image
        alt="Логотип"
        data-testid="image"
        src={EMPTY_IMAGE_SRC}
        width={40}
        height={40}
      />,
    );

    expect(screen.getByTestId("image")).toHaveAttribute("alt", "Логотип");
  });

  test("должен установить атрибут aria-hidden при пустом свойстве alt", () => {
    render(
      <Image
        alt=""
        data-testid="image"
        src={EMPTY_IMAGE_SRC}
        width={40}
        height={40}
      />,
    );

    expect(screen.getByTestId("image")).toHaveAttribute("alt", "");
    expect(screen.getByTestId("image")).toHaveAttribute("aria-hidden", "true");
  });

  test("должен сгенирировать placeholder при наличии свойства blurDataURL", () => {
    jest.useFakeTimers();

    render(
      <Image
        style={{ objectFit: "contain" }}
        alt="blurDataURL"
        data-testid="image"
        src={EMPTY_IMAGE_SRC}
        blurDataURL={EMPTY_IMAGE_SRC}
        width={40}
        height={40}
      />,
    );

    act(() => {
      jest.runAllTimers();
    });

    expect(screen.getByTestId("image")).toHaveStyle({ objectFit: "contain" });

    jest.useRealTimers();
  });

  test("должен появиться placeholder при наличии свойства blurDataURL", () => {
    jest.useFakeTimers();

    render(
      <Image
        alt=""
        src={EMPTY_IMAGE_SRC}
        blurDataURL={EMPTY_IMAGE_SRC}
        width={40}
        height={40}
      />,
    );

    act(() => {
      jest.runAllTimers();
    });

    expect(screen.getByTestId("image-placeholder")).toBeInTheDocument();

    jest.useRealTimers();
  });

  test("должен исчезнуть placeholder после окончания анимации", () => {
    jest.useFakeTimers();

    render(
      <Image
        alt=""
        src={EMPTY_IMAGE_SRC}
        blurDataURL={EMPTY_IMAGE_SRC}
        width={40}
        height={40}
      />,
    );

    act(() => {
      jest.runAllTimers();
    });

    const p = screen.getByTestId("image-placeholder");

    fireEvent.animationEnd(p);

    expect(p).not.toBeInTheDocument();

    jest.useRealTimers();
  });

  test("placeholder не должен появиться если изображение загружено", async () => {
    jest.useFakeTimers();

    render(
      <Image
        alt=""
        data-testid="image"
        src={EMPTY_IMAGE_SRC}
        blurDataURL={EMPTY_IMAGE_SRC}
        width={40}
        height={40}
      />,
    );

    await act(async () => {
      fireEvent.load(screen.getByTestId("image"));
    });

    act(() => {
      jest.runAllTimers();
    });

    expect(screen.queryByTestId("image-placeholder")).toBeNull();

    jest.useRealTimers();
  });
});
