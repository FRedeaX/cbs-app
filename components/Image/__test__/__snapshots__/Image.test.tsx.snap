// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> должен вернуть полный шаблон компонента (snapshot) 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1fasnog"
  >
    <img
      alt=""
      aria-hidden="true"
      class="image"
      data-nimg="1"
      decoding="async"
      height="40"
      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
      style="color: transparent;"
      width="40"
    />
  </div>
</DocumentFragment>
`;
